### PR TITLE
Streamparser and buffering fixes

### PR DESCRIPTION
**Pull request by kaidokert**
## Summary by Sourcery

Enable robust streaming JSON parsing by removing fixed escape reservations, adding buffer compaction support in StreamBuffer and StreamParser, refactoring position tracking APIs, and bolstering tests for low-memory and stress scenarios

New Features:
- Add StreamBuffer::compact_from method to slide unprocessed data and free buffer space
- Implement From<&str> conversion for JSON string wrapper

Enhancements:
- Remove static escape_reserve and allow full-buffer unescaped byte appends
- Refactor StreamParser to trigger buffer compaction on overflow and update parser state positions
- Simplify ContentRange API to track content_start for strings, keys, and numbers
- Adjust parsing logic to refill only when buffer is empty and integrate compaction for continuous streaming

Tests:
- Add comprehensive tests for buffer compaction behavior and edge cases
- Introduce stress tests for minimal buffer sizes, escape sequences, and varied read chunk scenarios

## Summary by CodeRabbit

* **New Features**
  * Added new methods for efficient buffer compaction and improved buffer management in streaming JSON parsing.
  * Introduced new trait implementations for easier string handling.
  * Enhanced streaming parser with robust handling of partial tokens and escape sequences.
  * Added comprehensive stress tests and new test utilities for validating parser behavior under various buffer and chunk size scenarios.

* **Refactor**
  * Simplified and clarified internal buffer and content boundary calculations.
  * Improved naming and separation of responsibilities in parser logic.

* **Bug Fixes**
  * Improved error handling for cases where buffer compaction discards necessary data.

* **Tests**
  * Added new targeted and stress tests to ensure correct parsing with small buffers and chunked input.

### CODE REVIEW

File: picojson/src/shared.rs [hunk 1]
```diff
@@ -124,59 +124,44 @@ pub trait PullParser {
...
+    /// * `current_pos` - Current parser position (after opening quote was processed)
     ///
     /// # Returns
-    /// Position that includes the first digit of the number
-    pub fn number_start_from_current(current_pos: usize) -> usize {
-        current_pos.saturating_sub(1) // Back up to include first digit
+    /// Position where string/key content begins (after the opening quote)
+    pub fn content_start_from_current(current_pos: usize) -> usize {
+        current_pos // Content starts at current position (after quote)
     }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/shared.rs]: This is done

File: picojson/src/stream_parser.rs [hunk 2]
```diff
@@ -426,10 +446,153 @@
...
+    /// Update parser state positions after buffer compaction
+    fn update_positions_after_compaction(&mut self, offset: usize) -> Result<(), ParseError> {
+        // Check for positions that would be discarded and need escape mode
+        // CRITICAL: Position 0 is never discarded, regardless of offset
+        let needs_escape_mode = match &self.parser_state.state {
+            crate::shared::State::Key(pos) if *pos > 0 && *pos < offset => Some((*pos, true)), // true = is_key
+            crate::shared::State::String(pos) if *pos > 0 && *pos < offset => Some((*pos, false)), // false = is_string
+            crate::shared::State::Number(pos) if *pos > 0 && *pos < offset => {
+                return Err(ParseError::ScratchBufferFull);
+            }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/stream_parser.rs]: done
